### PR TITLE
Improve performance of FirstValue, LastValue, Lead, Lag functions with IGNORE NULLS.

### PR DESCRIPTION
**Pull request by willmostly**
The current implementations rescan each window, potentially leading to O(N**2) performance. These implementations should only scan the new rows in each window, giving O(N) complexity.

### CODE REVIEW

**Comment by kokosing** [comment]
@willmostly Are you going to work on this? Or is it just FYI, so one can take this and continue to work on it?

**Comment by willmostly** [comment]
@kokosing I will work on it, unless it needs to be taken over for expediency.

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java [hunk 1]
```diff
@@ -27,11 +27,21 @@
 {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: commit message is too long:
```
Speed up some window functions with IGNORE NULLS

This commit improves perfomance of  FirstValue, LastValue, Lead, Lag 
window functions when they are used with INGORE NULLS clause.
```

File: presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java [hunk 2]
```diff
@@ -26,20 +26,31 @@
 @WindowFunctionSignature(name = "lag", typeVariable = "T", returnType = "T", argumentTypes = "T")
 @WindowFunctionSignature(name = "lag", typeVariable = "T", returnType = "T", argumentTypes = {"T", "bigint"})
 @WindowFunctionSignature(name = "lag", typeVariable = "T", returnType = "T", argumentTypes = {"T", "bigint", "T"})
+@WindowFunctionSignature(name = "lag", typeVariable = "T", returnType = "T", argumentTypes = {"bigint", "T", "boolean"})
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java]: It looks irrelevant to this commit.

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java [hunk 3]
```diff
@@ -27,11 +27,21 @@
...
     private final int argumentChannel;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: It would be nice to have some benchmark that measure the speedup? Did you do some performance testing?

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java [hunk 4]
```diff
@@ -41,24 +51,26 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
-                output.appendNull();
-                return;
-            }
+            returnInd = searchWindowAndUpdateIndices(frameStart, frameEnd);
+        }
+        else {
+            returnInd = frameStart;
         }
+        windowIndex.appendTo(argumentChannel, returnInd, output);
+    }
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: I would extract this if block as method so you don't need to have uninitialized variable

File: presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java [hunk 5]
```diff
@@ -26,20 +26,31 @@
...
+        lastNonNull = 0;
+    }
 
     public LagFunction(List<Integer> argumentChannels, boolean ignoreNulls)
     {
         this.valueChannel = argumentChannels.get(0);
         this.offsetChannel = (argumentChannels.size() > 1) ? argumentChannels.get(1) : -1;
         this.defaultChannel = (argumentChannels.size() > 2) ? argumentChannels.get(2) : -1;
         this.ignoreNulls = ignoreNulls;
+        if (this.ignoreNulls) {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java]: this block is not needed, `lastNonNull ` defaults to 0

File: presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java [hunk 6]
```diff
@@ -42,23 +54,33 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
             return;
         }
 
-        int valuePosition = frameEnd;
-
+        int returnInd;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java]: same comment as before

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java (hunk 3 above)
Comment (willmostly) [reply presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: I did some ad-hoc testing to show that it only checks the new values in each window, where the previous implementation rechecks the window for each row. I didn't do anything at scale personally, although there have been some significant performance hits observed with real data. What would you suggest?

Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: There is io.prestosql.operator.BenchmarkWindowOperator that benchmarks `ROW_NUMBER` function. Based on that we could write benchmarks for functions that you are improving.

WDYT?

File: presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java (hunk 2 above)
Comment (willmostly) [reply presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java]: Yep, mistake. Removed.

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java (hunk 4 above)
Comment (willmostly) [reply presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: Refactored to avoid uninitialized returnInd (didn't introduce a method though).

File: presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java (hunk 6 above)
Comment (willmostly) [reply presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java]: Same resolution as before :)

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java [hunk 7]
```diff
@@ -27,11 +27,21 @@
 {
     private final int argumentChannel;
     private final boolean ignoreNulls;
+    private int lastNonNull;
+
+    @Override
+    public void reset()
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: move below ctor

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java [hunk 8]
```diff
@@ -27,11 +27,21 @@
...
+    {
+        lastNonNull = -1;
+    }
 
     public FirstValueFunction(List<Integer> argumentChannels, boolean ignoreNulls)
     {
         this.argumentChannel = getOnlyElement(argumentChannels);
         this.ignoreNulls = ignoreNulls;
+        if (this.ignoreNulls) {
+            this.lastNonNull = -1;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: reset() is unconditional, why do we have condition here?

maybe just make this field's initial value?

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java [hunk 9]
```diff
@@ -27,11 +27,21 @@
 {
     private final int argumentChannel;
     private final boolean ignoreNulls;
+    private int lastNonNull;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: lastNonNull -> firstNonNull

File: presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java [hunk 10]
```diff
@@ -41,24 +51,23 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
-                output.appendNull();
-                return;
-            }
+            returnInd = searchWindowAndUpdateIndices(frameStart, frameEnd);
         }
+        windowIndex.appendTo(argumentChannel, returnInd, output);
+    }
 
-        windowIndex.appendTo(argumentChannel, valuePosition, output);
+    private int searchWindowAndUpdateIndices(int frameStart, int frameEnd)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/FirstValueFunction.java]: I think that inlining this (back) improves readability.

File: presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java [hunk 11]
```diff
@@ -33,6 +33,13 @@
     private final int offsetChannel;
     private final int defaultChannel;
     private final boolean ignoreNulls;
+    private long lastNonNull;
+
+    @Override
+    public void reset()
+    {
+        lastNonNull = 0;
+    }
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java]: move below ctor

why FirstValueFunction uses -1 as "not set" and this uses 0?
i think 0 is not a good value for unset (to be verified)

File: presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java [hunk 12]
```diff
@@ -52,23 +59,15 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
-                            break;
-                        }
-                    }
-                    valuePosition--;
+            if (ignoreNulls && withinPartition(valuePosition, currentPosition) && offset > 0) {
+                if (windowIndex.isNull(valueChannel, toIntExact(valuePosition))) {
+                    valuePosition = lastNonNull;
+                }
+                else {
+                    lastNonNull = valuePosition;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/LagFunction.java]: I think it isn't correct

for `lag(3)` if you check 3 positions abck and this happens to be a non-null value, you do not know whether it's 3-non-null values back. It may be the previous non null (`lag(1)`)

do we have some tests that would excercise this scenario?

File: presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java [hunk 13]
```diff
@@ -27,11 +27,23 @@
...
     private final boolean ignoreNulls;
+    private int lastNonNull;
+    private int previousFrameEnd;
+
+    @Override
+    public void reset()
+    {
+        lastNonNull = -1;
+        previousFrameEnd = 0;
+    }
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java]: move below ctor

File: presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java [hunk 14]
```diff
@@ -42,23 +54,30 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
 
-            if (valuePosition < frameStart) {
-                output.appendNull();
-                return;
+    private int searchWindowAndUpdateIndices(int frameStart, int frameEnd)
+    {
+        if (previousFrameEnd < frameStart) {
+            previousFrameEnd = frameStart;
+        }
+        //This search assumes frameEnd is monotonically increasing.
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java]: verify instead of assuming

File: presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java [hunk 15]
```diff
@@ -42,23 +54,30 @@ public void processRow(BlockBuilder output, int frameStart, int frameEnd, int cu
...
-
-                valuePosition--;
-            }
+            returnInd = searchWindowAndUpdateIndices(frameStart, frameEnd);
+        }
+        windowIndex.appendTo(argumentChannel, returnInd, output);
+    }
 
-            if (valuePosition < frameStart) {
-                output.appendNull();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/LastValueFunction.java]: previously there was a case where we would `appendNull`

not sure yet under what conditions, but where is it now?

File: presto-main/src/main/java/io/prestosql/operator/window/LeadFunction.java [hunk 16]
```diff
@@ -33,6 +33,13 @@
     private final int offsetChannel;
     private final int defaultChannel;
     private final boolean ignoreNulls;
+    private long lastNonNull = -1;
+
+    @Override
+    public void reset()
+    {
+        lastNonNull = -1;
+    }
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/window/LeadFunction.java]: move below ctor

**Comment by martint** [comment]
@willmostly , are you still working on this?

**Comment by mosabua** [comment]
:wave: @willmostly  - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.
